xdsl/dialects/csl/csl_stencil.py

```diff
@@ -121,6 +141,237 @@ def __init__(
         )
 
 
+@irdl_op_definition
+class ApplyOp(IRDLOperation):
+    """
+    This operation combines a `csl_stencil.prefetch` (symmetric buffer communication across a given stencil shape)
+    with a `stencil.apply` (a stencil function plus parameters and applies the stencil function to the output temp).
+
+    As communication may be done in chunks, this operation provides two regions for computation:
+      - the `chunk_reduce` region to reduce a chunk of data received from several neighbours to one chunk of data.
+        this region is invoked once per communicated chunks and effectively acts as a loop body.
+        It uses `iter_arg` to concatenate the chunks
+      - the `post_process` region (invoked once when communication has finished) that takes the concatenated
+        chunk of the `chunk_reduce` region and applies any further processing here - for instance, it may handle
+        the computation of 'own' (non-communicated) or otherwise prefetched data
+
+    Further fields:
+      - `communicated_stencil` - the stencil to communicate (send and receive)
+      - `args`  - arguments to the stencil computation, may include other prefetched buffers
+      - `topo`  - as received from `csl_stencil.prefetch`/`dmp.swap`
+      - `swaps` - a set of neighbouring points in the stencil, whose value we wish to retain
+                  (note, these are not guaranteed to be lowered as true point-to-point communication, and
+                  redundant communication should be irgnored)
+
+    Function signatures:
+    Before lowering (from `csl_stencil.prefetch` and `stencil.apply`):
+        %pref = csl_stencil.prefetch(%communicated_stencil : stencil.Temp)
+        stencil.apply( ..some args.. , %communicated_stencil, ..some more args.., %pref)
+
+    After lowering:
+        op:             csl_stencil.apply(%communicated_stencil, %iter_arg, chunk_reduce_args..., post_process_args...)
+        chunk_reduce:   block_args(slice of type(%pref), %offset, %iter_arg, args...)
+        post_process:   block_args(%communicated_stencil, %iter_arg, args...)
+
+    Note, that %pref can be dropped (as communication is done by the op rather than before the op),
+    and that a new %iter_arg is required, an empty tensor which is filled by `chunk_reduce` and
+    consumed by `post_process`
+    """
+
+    name = "csl_stencil.apply"
+
+    communicated_stencil = operand_def(
+        stencil.TempType[Attribute] | memref.MemRefType[Attribute]
+    )
+
+    iter_arg = operand_def(TensorType[Attribute])
+
+    args = var_operand_def(Attribute)
+
+    chunk_reduce = region_def()
+    post_process = region_def()
+
+    swaps = prop_def(builtin.ArrayAttr[ExchangeDeclarationAttr])
+
+    topo = opt_prop_def(dmp.RankTopoAttr)
+
+    num_chunks = prop_def(IntegerAttr)
```

Comment:
There seem to be some undocumented attributes here still, can you add some info on what they do? (Maybe as a docstring directly attached to the attributes?

Reply:
Adding info on what they do